### DISCUSSION TITLE
I2S Full-duplex: Record and Play. Possible with the Pico?

### DISCUSSION THREAD

**Discussion by hexagon5un** [micropython/micropython#12473] [RP2040 / Pico] [answered]
Hi,

I'm working on a project with an AK4619 I2S DAC/ADC chip.  (It's really great!)  It's capable of doing 4 channels in and out simultaneously, as TDM streams or as multiple stereo in/out I2S streams, but I'm so far unable to make the Pi Pico do it in Micropython.

(The first challenge was implementing a master clock, but the PIOs made quick work of that!)

The fundamental problem in the Micropython implmentation seems to be that you can instantiate two I2S objects, but they're not synchronized, so the clocks fight.   

I see two solutions: 

1) Implement an I2S class that can take a clock from another instance.  This would also allow two simultaneous stereo streams, which this I2S codec supports.  Or, of course, full duplex. 

2) Make the read function somehow not exclusive of the write function.  In fact, now that I say this, why are they exclusive in the first place?

Anyone who knows the codebase: Which if these approaches should be easier or more valuable?  (I need to get this working in the next month or so, but I'm willing to hack on the Micropython codebase afterwards if it'll be useful to other folks.)

Thanks! 

   Elliot.

**Comment by miketeachman** [comment, marked as answer]
I recommend modifying the I2S class to make it full duplex.  I think that will be easier than trying to synchronize the DMA channels between independent I2S instances.   Here are some implementation ideas:
- change the single SD pin to two pins, SDIN, SDOUT
- add a new mode to support full duplex, such as RXTX
- create two application buffers, one for TX, one for RX.  The split would be internal to the class.  No need to expose this implementation detail to the user API.
- create two DMA buffers, one for TX, one for RX
- keep the 2 chained DMA channels, but now the DMA interrupt both feeds and empties the DMA buffers
- modify the PIO code so that it handles full duplex.

For the master clock you might also consider using the PWM machine class.  That might be easier than using a PIO?

note:  I'm the author of the MicroPython I2S C code.

**Reply by hexagon5un** [reply]
Thanks very much for the advice!  I'll have to start digging into the I2S implementation seriously, then.  

Am I hopelessly naive to think that since there's already most of the functionality implemented, that it'll just be a matter of making sure that they're both "turned on" at once?  

Do you have any idea how many people are using the I2S now?  Changing SD to SDIN and SDOUT is a breaking change.  

Using the PIO for MCLK was easy enough, but if the maximum speed on the PWM peripheral is fast enough, why not?  I'll look into it.  I don't know how scarce the PIOs are in Micropython implementations yet, and wouldn't want to burn one when a PWM would do.

(And thanks for your wonderful example code on using the I2S in Micropython.  It was very helpful when I was trying to get my proof-of-concepts up and running.)

**Reply by miketeachman** [reply]
Implementing full duplex will mostly be a matter of understanding the existing code and making incremental modifications.   You'll need reasonable C experience and will need to spend some time understanding the Pico SDK.    You raise a good point about the breaking change.  I2S has been in MicroPython for over two years and many people have used it, so, any changes should likely be backwards compatible.  I think that decision is up to the maintainers.  Maybe the 2nd serial data pin could be named something like SD2.  For full duplex, SD could be the input pin, and SD2 could be the output pin.   I'm sure someone can come up with something better.

**Reply by hexagon5un** [reply]
This is going to be a challenge for sure, and I'm not 100% sure that what I hack together is going to be suitable for upstreaming. I really need to get it working first, and then I can think about integration in late November.  :)

That said, I would do SD_RX and SD_TX internally?  Code could allocate the pins dependent on the mode, which would solve the backwards compat issue.  SD_RX -> SD when i2s_mode == RX, etc.  

I can't figure out how to get one PIO to do _both_ the RX and TX, though.  I might need 2 PIOs, which means synching them ...  This gets hairy with the whole check-for-available-PIO thing in the codebase.  

I'm halfway comforatble with C in general, but I've got a whole lot of learning how to integrate the C back into Micropython. At this point that's a hard requirement, so I will learn!

**Comment by A622266** [comment]
Hi Elliott
I'm interested in this work.  I posted previously on the topic of I2S and Micropython:
https://github.com/orgs/micropython/discussions/9639

If you read through the whole thread, you'll see that I thought I could get away with half-duplex, but I discovered that there are many limitations, including that you cannot use products like AK4619.  Full Duplex is a very important feature for Micropython I2S support, and will allow the community to leverage many advanced audio codec products.

BTW, what can you do with a Pico and I2S?  An ESP32S3 or Teensy are more capable for DSP processing. Or is it just a controller, and you use a specialized DSP as companion to AK4619?

Regards
Andy

**Reply by hexagon5un** [reply]
I'm not really doing much / any DSP.  The project is a low-res waveform generator and oscilloscope combo, and I need to get them both running simultaneously off of the Pico.  

On the signal generator side, I'm happy with DDS, and then whatever else people can implement -- and I'm dedicating a whole Pico core to this thread.  How far one can push this is part of a challenge.  

On the scope side, it's really just taking the data in both channels and scaling / blitting out to a screen.  And making it look good.  :)
